src/SecTester.Scan/Scan.cs

```diff
@@ -1,58 +1,115 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
 using SecTester.Scan.Models;
 
 namespace SecTester.Scan;
 
-public class Scan : IDisposable
+public class Scan : IAsyncDisposable
 {
+  private static readonly TimeSpan DefaultPollingInterval = TimeSpan.FromSeconds(5);
+
   private static readonly IReadOnlyCollection<ScanStatus> ActiveStatuses =
-    new[]
-    {
-      ScanStatus.Pending, ScanStatus.Running, ScanStatus.Queued
-    };
+    new[] { ScanStatus.Pending, ScanStatus.Running, ScanStatus.Queued };
 
   private static readonly IReadOnlyCollection<ScanStatus> DoneStatuses =
-    new[]
-    {
-      ScanStatus.Disrupted, ScanStatus.Done, ScanStatus.Failed, ScanStatus.Stopped
-    };
+    new[] { ScanStatus.Disrupted, ScanStatus.Done, ScanStatus.Failed, ScanStatus.Stopped };
 
   private readonly ScanOptions _options;
-
+  private readonly ILogger _logger;
   private readonly Scans _scans;
-  private readonly ScanState _scanState = new(ScanStatus.Pending);
+  private volatile ScanState _scanState = new(ScanStatus.Pending);
 
-  public Scan(Scans scans, ScanOptions options)
+  public string Id { get; init; }
+
+  public Scan(string id, Scans scans, ILogger logger, ScanOptions options)
   {
+    Id = id ?? throw new ArgumentNullException(nameof(id));
     _scans = scans ?? throw new ArgumentNullException(nameof(scans));
     _options = options ?? throw new ArgumentNullException(nameof(options));
+    _logger = logger ?? throw new ArgumentNullException(nameof(logger));
   }
 
   public bool Active => ActiveStatuses.Contains(_scanState.Status);
 
   public bool Done => DoneStatuses.Contains(_scanState.Status);
 
-  public void Dispose()
+  public async ValueTask DisposeAsync()
   {
+    try
+    {
+      await RefreshState().ConfigureAwait(false);
+
+      if (!Active)
+      {
+        await _scans.DeleteScan(Id).ConfigureAwait(false);
+      }
+    }
+    catch
+    {
+      // noop
+    }
+
     GC.SuppressFinalize(this);
   }
 
-  public Task<IEnumerable<Issue>> Issues(CancellationToken cancellationToken = default)
+
+  public Task<IEnumerable<Issue>> Issues()
   {
-    throw new NotImplementedException();
+    return _scans.ListIssues(Id);
   }
 
   public Task Stop()
   {
-    throw new NotImplementedException();
+    return _scans.StopScan(Id);
```

Comment:
@derevnjuk What do you think if the `Stop` itself may be called concurrently? Thus having the latest state from `App` is worthless because of race conditions and may also lead to an unwanted call of `Scans.StopScan`.

Reply:
The comment above gives you a precise answer to this question. To get into the race conditions using a lock, someone should refresh the state to done-like status when another thread is between `Active` and `Stop` expressions. That is why you have to silently handle such exception. But it is not possible to get into this bug just calling `Stop` concurently on the condition that one of thread accuare a lock.